crates/chaintsn/src/transition.rs

```diff
@@ -626,4 +635,82 @@ mod tests {
     }
+
+    #[test]
+    fn test_process_l1_block_with_duplicate_withdrawal_fulfillment() {
+        let mut arb = ArbitraryGenerator::new();
+
+        // Setup chainstate with a deposit in dispatched state
+        let mut chs: Chainstate = arb.generate();
+        let mut deposit_table = DepositsTable::new_empty();
+        deposit_table.create_next_deposit(
+            arb.generate(),
+            vec![0, 1, 2],
+            BitcoinAmount::from_int_btc(10),
+        );
+        let mut deposit = deposit_table.get_deposit_mut(0).expect("should exist");
+        deposit.set_state(DepositState::Dispatched(DispatchedState::new(
+            DispatchCommand::new(vec![]),
+            0,
+            100,
+        )));
+        *chs.deposits_table_mut() = deposit_table;
+
+        let params = gen_params();
+
+        // withdrawal fulfillment op that will be seen twice
+        let withdrawal_fulfillment_protocol_op =
+            ProtocolOperation::WithdrawalFulfillment(WithdrawalFulfillmentInfo {
+                deposit_idx: 0,
+                operator_idx: 0,
+                amt: BitcoinAmount::from_int_btc(10),
+                txid: Buf32::zero(),
+            });
+
+        // send protocol op for first time. Should set deposit to fulfilled state
+        let block_manifest = L1BlockManifest::new(
+            arb.generate(),
+            None,
+            vec![L1Tx::new(
+                arb.generate(),
+                arb.generate(),
+                vec![withdrawal_fulfillment_protocol_op.clone()],
+            )],
+            chs.cur_epoch(),
+            chs.l1_view().safe_height() + 1,
+        );
+
+        let mut state_cache = StateCache::new(chs.clone());
+        let result = process_l1_block(&mut state_cache, &block_manifest, params.rollup());
+        assert!(result.is_ok(), "should succeed");
+        let new_deposit_state = state_cache
+            .state()
+            .deposits_table()
+            .get_deposit(0)
+            .expect("should exist")
+            .deposit_state();
+        assert!(
+            matches!(new_deposit_state, DepositState::Fulfilled(_)),
+            "should convert state to fulfilled"
+        );
+
+        // Duplicate withdrawal fulfillment seen. Should be ignored and chainstate not updated
+        let block_manifest = L1BlockManifest::new(
+            arb.generate(),
+            None,
+            vec![L1Tx::new(
+                arb.generate(),
+                arb.generate(),
+                vec![withdrawal_fulfillment_protocol_op],
+            )],
+            chs.cur_epoch(),
+            chs.l1_view().safe_height() + 1,
+        );
+
+        let chs = state_cache.state();
+        let mut state_cache = StateCache::new(chs.clone());
+        let result = process_l1_block(&mut state_cache, &block_manifest, params.rollup());
+        assert!(result.is_ok(), "should succeed");
+        assert_eq!(state_cache.state(), chs, "should not update chainstate");
```

Comment:
Should be `test: invoke process_l1_block` and `test: chainstate changed unexpectedly`